### PR TITLE
Load time optimizations

### PR DESCRIPTION
**Pull request by MarioSteinitz**
With this merge request, we targeted load time optimizations. The changes we did are as follows:

  * Add a config.yml setting `mode`. If this setting has been set to `production`, all JavaScript and CSS will be served minified.
  * Add the `bust` query string parameter to all JavaScript files loaded using require.js. (We set our expiry headers to half a year for js, css, fonts and images. Having a bust parameter here assures they will be updated with at least every version switch.)
  * Add an unminified version of every third-party JavaScript used in TBG to the /public/js folders.
  * Remove all minified JavaScript from the repository and add it to the .gitignore.
  * *Add Grunt to minify CSS and JS.* In order to using Grunt, Node.js and NPM must be installed. Then on the cli run `npm install` and `node_modules/.bin/grunt` within the TBG root folder. The Grunt task will remove all minified JavaScript and CSS within the `/public/js`, `public/css`, `modules/*/public/js`, `modules/*/public/css`, `themes/*/js` and and `themes/*/css` folders and generate new minified versions.
  * Add a small placeholder image to ui.inc.php's `image_tag` function which will be resolved to the real images using JavaScript, after domReady was triggered. So other TBG JS initializers will be ready set before all images are fully loaded.

Please let me know what you think.

### CODE REVIEW

**Comment by azaghal** [comment]
Personally, I would really prefer not to have to install an additional tool in production in order to serve minified JS/CSS, especially given that often one needs to install up-to-date Node.js etc (e.g. usually you what is shipped by distros is too old). Then again, this could be achieved by making sure that at least the releases include minified JS/CSS.

**Comment by MarioSteinitz** [comment]
Well, removing the minified sources is a suggestion. For sure Grunt can be used as development tool only for generating the minified files, once they were changed.
In our own TBG installation we added it to gitignore to keeping the repo clean. (We do have all these tools pre-installed on all our servers. So running a Grunt task on the target server is not much of a hassle for us.)

**Comment by zegenie** [comment]
I don't personally mind adding other tools to the tool chain, but I'll have
to look through the PR before I vote 😊

**Comment by MarioSteinitz** [comment]
IMHO it's a matter of taste whether to add the minified files or not. Our company's rules is to not add stuff to version control that can be generated or downloaded using package managers. Actually like the vendor files of composer packages. Such stuff will be included in release packages only.

But composer dependency management reminds me of another topic I'd like to ask before we work on it.-  Is there any active channel to talking about such stuff with the core dev team? Irc, Slack, Mattermost, Skype, ...?

**Comment by azaghal** [comment]
There is gitter (https://gitter.im/thebuggenie/development and https://gitter.im/thebuggenie/general).

As I mentioned, I would really like the releases to include the minified stuff out-of-the-box (@zegenie - would this be an option?). Otherwise for development side I'm fine - plus it should be relatively easy to add to dev machine - there's included Vagrantfile and Ansible playbooks :)

File: core/framework/Context.php [hunk 1]
```diff
@@ -2729,4 +2729,16 @@ public static function setDebugMode($value = true)
         self::$_debug_mode = $value;
     }
 
+    /**
+     * Whether to serve minified resource files
+     *
+     * @return bool
+     *   true, if resouce files shall be delivered minified, false otherwise.
+     */
+    public static function isServeMinified()
```
Comment (zegenie) [inline core/framework/Context.php]: I don't know what "is serve minified" means :)

File: core/config/settings.yml [hunk 2]
```diff
@@ -1,3 +1,4 @@
 core:
     debug: false
     log_file: ~
+    mode: production
```
Comment (zegenie) [inline core/config/settings.yml]: in addition, I think we should introduce a specific flag for serving minified content, so `minified_assets` or something similar

File: public/js/thebuggenie.js [hunk 3]
```diff
@@ -55,6 +55,10 @@ define(['thebuggenie/tools', 'thebuggenie/tbg', 'domReady', 'jquery', 'mention']
                     TBG.Main.Profile.clearPopupsAndButtons();
                     e.stopPropagation();
                 });
+                jQuery("img[data-src]:not([data-src-processed])").each(function(){
+                    var $img = jQuery(this);
+                    $img.attr('src', $img.data('src')).data('src-processed', true);
+                });
```
Comment (MarioSteinitz) [inline public/js/thebuggenie.js]: When images are loaded via ajax, then the images come as is in src. Asynchronous image loading didn't make much sense to us here.

File: core/framework/Context.php (hunk 1 above)
Comment (MarioSteinitz) [inline core/framework/Context.php]: Shall we rename it? Which name do you suggest?

**Pull request merged**